volume/drivers/coprhd/coprhd.go

```diff
@@ -217,40 +408,82 @@ func (d *driver) Snapshot(
 	volumeID string,
 	readonly bool,
 	locator *api.VolumeLocator) (string, error) {
-	return "", nil
+	return "", volume.ErrNotSupported
 }
 
 func (v *driver) Status() [][2]string {
 	return [][2]string{}
 }
 
-// getAuthSession returns an authenticated API Session
-func (d *driver) getAuthSession() (session *napping.Session, err error) {
-	e := ApiError{}
-
-	s := napping.Session{
-		Userinfo: d.creds,
-		Client:   d.httpClient,
+func (d *driver) Inspect(volumeIDs []string) ([]*api.Volume, error) {
+	vols, err := d.DefaultEnumerator.Inspect(volumeIDs)
+	if err != nil {
+		return nil, err
 	}
 
-	url := d.url + loginUri
-
-	resp, err := s.Get(url, nil, nil, &e)
+	for i, v := range vols {
+		proto := d.itr.Protocol
+		dev := ""
+
+		switch proto {
+		case coprhd.InitiatorTypeScaleIO:
+			dev, err = d.waitAttachDevice(v.Id, time.Second*1)
+			if err == nil {
+				vols[i].DevicePath = dev
+			}
+		default:
+		}
+	}
+	return vols, nil
+}
 
+func (d *driver) Format(volumeID string) error {
+	v, err := d.GetVol(volumeID)
 	if err != nil {
-		return
+		return fmt.Errorf("Volume is not attached")
 	}
 
-	token := resp.HttpResponse().Header.Get("X-SDS-AUTH-TOKEN")
+	fstype := v.Spec.Format.SimpleString()
+	devicePath := v.DevicePath
 
-	h := http.Header{}
-
-	h.Set("X-SDS-AUTH-TOKEN", token)
+	cmd := "/sbin/mkfs." + fstype
+	o, err := exec.Command(cmd, devicePath).Output()
+	if err != nil {
+		dlog.Warnf("Failed to run command %v %v: %v", cmd, devicePath, o)
+		return err
+	}
+	v.Format = v.Spec.Format
+	err = d.UpdateVol(v)
+	return err
+}
 
-	session = &napping.Session{
-		Client: d.httpClient,
-		Header: &h,
+func (d *driver) waitAttachDevice(volumeID string, to time.Duration) (string, error) {
+	_, err := d.GetVol(volumeID)
+	if err != nil {
+		return "", fmt.Errorf("Volume is not attached")
 	}
 
-	return
+	timeout := time.After(to)
+	timer := time.Tick(time.Millisecond * 100)
+
+	proto := d.itr.Protocol
+	dev := ""
+
+	for {
+		switch proto {
```

Comment:
for the moment scaleio is the only support protocol...but this will change and the device check is protocol specific.